### PR TITLE
Add zImage parser for ARM

### PR DESCRIPTION
**Pull request by rjoleary**
This will be used for arm kexec.

Signed-off-by: Ryan O'Leary <ryanoleary@google.com>

### CODE REVIEW

**Comment by rjoleary** [comment]
See #401

File: pkg/zimage/zimage.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+//
+// The layout is defined in Linux:
+//     arch/arm/boot/compressed/vmlinux.lds.S
+type TableEntry struct {
+	Tag  uint32
+	Data []uint32
+}
+
+// Parse a ZImage from a file.
+func Parse(f *os.File) (*ZImage, error) {
```
Comment (n-canter) [inline pkg/zimage/zimage.go]: I think io.ReadSeeker is better then os.File
Then you will be able to create simple tests.

File: pkg/zimage/zimage.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	TableMagic uint32 `offset:"0x10"`
+	TableAddr  uint32 `offset:"0x14"`
+}
+
+// Table is an extension to Header. A zImage may have 0 or more entries.
+//
+// The layout is defined in Linux:
+//     arch/arm/boot/compressed/vmlinux.lds.S
+type TableEntry struct {
+	Tag  uint32
```
Comment (n-canter) [inline pkg/zimage/zimage.go]: can it be useful to create type for Tag and Data? like
```
type Tag uint32
type Data uint32
```

File: pkg/zimage/zimage.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	if z.Header.Endianess != Endianess {
+		return z, fmt.Errorf("unsupported zImage endianess, expected little")
+	}
+	if z.Header.End < z.Header.Start {
+		return z, fmt.Errorf("invalid zImage, end is less than start, %d < %d",
+			z.Header.End, z.Header.Start)
+	}
+
+	// Parse the table.
+	if z.Header.TableMagic == TableMagic {
```
Comment (n-canter) [inline pkg/zimage/zimage.go]:
```
if z.Header.TableMagic != TableMagic {
    return ...
}
addr := z.header.TableAddr
...
```

File: pkg/zimage/zimage.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		addr := z.Header.TableAddr
+		for addr != 0 {
+			if _, err := f.Seek(int64(addr), os.SEEK_SET); err != nil {
```
Comment (n-canter) [inline pkg/zimage/zimage.go]: SEEK_SET deprecated, use io.SeekStart

File: pkg/zimage/zimage.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	Header Header
+	Table  []TableEntry
+}
+
+// Header appears near the beginning of the zImage.
+//
+// The layout is defined in Linux:
+//     arch/arm/boot/compressed/head.S
+type Header struct {
+	Magic      uint32 `offset:"0x0"`
```
Comment (n-canter) [inline pkg/zimage/zimage.go]: Can't see where you use tag 'offset'? Are you going to use reflection at some point?

File: pkg/zimage/zimage.go (hunk 1 above)
Comment (rjoleary) [reply pkg/zimage/zimage.go]: Done!

File: pkg/zimage/zimage.go (hunk 3 above)
Comment (rjoleary) [reply pkg/zimage/zimage.go]: Done!

File: pkg/zimage/zimage.go (hunk 4 above)
Comment (rjoleary) [reply pkg/zimage/zimage.go]: Thanks! I keep forgetting.

File: pkg/zimage/zimage.go (hunk 5 above)
Comment (rjoleary) [reply pkg/zimage/zimage.go]: Good point, I removed it. The "pkg/bzimage/header.go" also does this -- not too sure why.

**Pull request merged**